db/write_batch_internal.h

```diff
@@ -260,4 +260,173 @@ class LocalSavePoint {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status DeleteCF(uint32_t cf, const Slice& key) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status SingleDeleteCF(uint32_t cf, const Slice& key) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status DeleteRangeCF(uint32_t cf, const Slice& begin_key,
+                       const Slice&) override {
+    Status s = AssignTimestamp(cf, begin_key);
+    ++idx_;
+    return s;
+  }
+
+  Status MergeCF(uint32_t cf, const Slice& key, const Slice&) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status PutBlobIndexCF(uint32_t cf, const Slice& key, const Slice&) override {
+    Status s = AssignTimestamp(cf, key);
+    ++idx_;
+    return s;
+  }
+
+  Status MarkBeginPrepare(bool) override { return Status::OK(); }
+
+  Status MarkEndPrepare(const Slice&) override { return Status::OK(); }
+
+  Status MarkCommit(const Slice&) override { return Status::OK(); }
+
+  Status MarkRollback(const Slice&) override { return Status::OK(); }
+
+ protected:
+  virtual Status AssignTimestamp(uint32_t cf, const Slice& key) = 0;
+  // No copy or move.
+  TimestampAssignerBase(const TimestampAssignerBase&) = delete;
+  TimestampAssignerBase(TimestampAssignerBase&&) = delete;
+  TimestampAssignerBase& operator=(const TimestampAssignerBase&) = delete;
+  TimestampAssignerBase&& operator=(TimestampAssignerBase&&) = delete;
+
+  WriteBatch::ProtectionInfo* const prot_info_ = nullptr;
+  const Checker checker_{};
+  size_t idx_ = 0;
+};
+
+template <typename Checker>
+class SimpleListTimestampAssigner : public TimestampAssignerBase<Checker> {
+ public:
+  explicit SimpleListTimestampAssigner(WriteBatch::ProtectionInfo* prot_info,
+                                       Checker checker,
+                                       const std::vector<Slice>& timestamps)
+      : TimestampAssignerBase<Checker>(prot_info, std::move(checker)),
+        timestamps_(timestamps) {}
+
+  ~SimpleListTimestampAssigner() override {}
+
+ private:
+  using TimestampAssignerBase<Checker>::prot_info_;
+  using TimestampAssignerBase<Checker>::checker_;
+  using TimestampAssignerBase<Checker>::idx_;
+
+  Status AssignTimestamp(uint32_t cf, const Slice& key) override {
+    if (idx_ >= timestamps_.size()) {
+      return Status::InvalidArgument("Need more timestamps for the assignment");
+    }
+    const Slice& ts = timestamps_[idx_];
+    size_t ts_sz = 0;
+    if (checker_) {
+      const Status s = checker_(cf, ts_sz);
+      if (!s.ok()) {
+        return s;
+      } else if (ts_sz != ts.size()) {
+        return Status::InvalidArgument("Timestamp size mismatch");
+      }
+    } else {
+      ts_sz = ts.size();
+    }
+    if (0 == ts_sz) {
+      // This key does not have timestamp, so skip.
+      return Status::OK();
+    }
```

Comment:
How about moving this to a helper in the base class which would take `ts` (in this class, `timestamps_[idx]`, in the other one, `timestamp_`) as a parameter?

Similarly, the snippet that deals with `prot_info_` and the one that computes the address and calls `memcpy` could also be helper methods I think.

This would probably also eliminate the need for the `using` declarations in the derived classes.

Reply:
Done.

Reply:
Done